File: lib/ts/src/internal/ParseRequestMessage.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { Serializer } from '../Serializer.js';
import { TelepactSchema } from '../TelepactSchema.js';
import { Message } from '../Message.js';
import { BinaryEncoderUnavailableError } from '../internal/binary/BinaryEncoderUnavailableError.js';
import { BinaryEncodingMissing } from '../internal/binary/BinaryEncodingMissing.js';
import { InvalidMessage } from '../internal/validation/InvalidMessage.js';
import { InvalidMessageBody } from '../internal/validation/InvalidMessageBody.js';
import { TelepactError } from '../TelepactError.js';

export function parseRequestMessage(
    requestMessageBytes: Uint8Array,
    serializer: Serializer,
    telepactSchema: TelepactSchema,
    onError: (error: TelepactError) => void,
): Message {
    try {
        return serializer.deserialize(requestMessageBytes);
    } catch (e) {
        onError(new TelepactError('telepact request parsing failed while decoding the incoming message', 'parse', e));

        let reason: string;
        if (e instanceof BinaryEncoderUnavailableError) {
            reason = 'IncompatibleBinaryEncoding';
        } else if (e instanceof BinaryEncodingMissing) {
            reason = 'BinaryDecodeFailure';
        } else if (e instanceof InvalidMessage) {
            reason = 'ExpectedJsonArrayOfTwoObjects';
        } else if (e instanceof InvalidMessageBody) {
            reason = 'ExpectedJsonArrayOfAnObjectAndAnObjectOfOneObject';
        } else {
            reason = 'ExpectedJsonArrayOfTwoObjects';
        }

        return new Message({ _parseFailures: [{ [reason]: {} }] }, { _unknown: {} });
    }
}

File: lib/go/internal/ParseRequestMessage.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package internal

import (
	"errors"
	"fmt"

	"github.com/telepact/telepact/lib/go/internal/binary"
	"github.com/telepact/telepact/lib/go/internal/types"
)

func ParseRequestMessage(
	requestMessageBytes []byte,
	deserialize func([]byte) (ServerMessage, error),
	schema SchemaAccessor,
	onError func(error),
) (ServerMessage, error) {
	_ = schema

	message, err := deserialize(requestMessageBytes)
	if err == nil {
		return message, nil
	}

	invokeOnError(onError, fmt.Errorf("telepact request parsing failed while decoding the incoming message: %w", err))

	reason := "ExpectedJsonArrayOfTwoObjects"

	var unavailableErr binary.BinaryEncoderUnavailableError
	if errors.As(err, &unavailableErr) {
		reason = "IncompatibleBinaryEncoding"
	} else {
		var missingErr *binary.BinaryEncodingMissing
		if errors.As(err, &missingErr) {
			reason = "BinaryDecodeFailure"
		} else {
			var invalidMessageErr *types.InvalidMessage
			if errors.As(err, &invalidMessageErr) {
				reason = "ExpectedJsonArrayOfTwoObjects"
			} else {
				var invalidBodyErr *types.InvalidMessageBody
				if errors.As(err, &invalidBodyErr) {
					reason = "ExpectedJsonArrayOfAnObjectAndAnObjectOfOneObject"
				}
			}
		}
	}

	parseHeaders := map[string]any{
		"_parseFailures": []any{map[string]any{reason: map[string]any{}}},
	}
	parseBody := map[string]any{
		"_unknown": map[string]any{},
	}

	return ServerMessage{Headers: parseHeaders, Body: parseBody}, nil
}
